nomenklatura/resolver/resolver.py

```diff
@@ -1,90 +1,320 @@
+        self._transaction = self._conn.begin()
+
+    def commit(self) -> None:
+        if self._transaction is None or self._conn is None:
+            raise RuntimeError("No transaction to commit.")
+        self._transaction.commit()
+        self._transaction = None
+        self._conn.close()
+        self._conn = None
+
+    def rollback(self, force: bool = False) -> None:
+        if self._transaction is not None:
+            self._transaction.rollback()
+            self._transaction = None
+        else:
+            if not force:
+                raise RuntimeError("No transaction to rollback.")
+        if self._conn is not None:
+            self._conn.close()
+            self._conn = None
+        else:
+            if not force:
+                raise RuntimeError("No connection to close.")
+
+    def _get_connection(self) -> Connection:
+        if self._transaction is None or self._conn is None:
+            raise RuntimeError("No transaction in progress.")
+        return self._conn
 
     def get_linker(self) -> Linker[CE]:
         """Return a linker object that can be used to resolve entities.
         This is less memory-consuming than the full resolver object.
         """
-        entities: Dict[Identifier, Set[Identifier]] = {}
-        for node in self.nodes.keys():
-            if node in entities:
-                continue
-            connected = self._traverse(node, set())
-            for c in connected:
-                entities[c] = connected
-        return Linker(entities)
+        clusters: Dict[Identifier, Set[Identifier]] = {}
+        stmt = select(self._table)
+        stmt = stmt.where(self._table.c.judgement == Judgement.POSITIVE.value)
+        stmt = stmt.where(self._table.c.deleted_at.is_(None))
+        for row in self._get_connection().execute(stmt).fetchall():
+            edge = Edge.from_dict(row._mapping)
+            cluster = clusters.get(edge.target, set())
+            cluster.update(clusters.get(edge.source, [edge.source]))
+            cluster.add(edge.target)
+            for node in cluster:
+                clusters[node] = cluster
+        log.info("Loaded %s clusters from: %s", len(clusters), self)
+        return Linker(clusters)
+
+    def warm_linker(self) -> None:
+        """Preload the linker cache."""
+        self._linker = self.get_linker()
 
     def get_edge(self, left_id: StrIdent, right_id: StrIdent) -> Optional[Edge]:
+        """Get an edge matching the given keys in any direction, if it exists."""
+
         key = Identifier.pair(left_id, right_id)
-        return self.edges.get(key)
+        stmt = self._table.select()
+        stmt = stmt.where(self._table.c.target == key[0].id)
+        stmt = stmt.where(self._table.c.source == key[1].id)
+        stmt = stmt.where(self._table.c.deleted_at.is_(None))
+        res = self._get_connection().execute(stmt).fetchone()
+        if res is None:
+            return None
+        return Edge.from_dict(res._mapping)
```

Comment:
is there a public way of doing this?

Reply:
this is the documented interface https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Row._mapping